### PR TITLE
feat: adf_to_text rich rendering — ordered lists, marks, tables, blockquote, rule, hardBreak (#202)

### LINKED ISSUE #202: feat: adf_to_text should render ordered lists numerically (and handle tables/marks/blockquotes)
## Problem

`src/adf.rs::adf_to_text()` (used by `jr issue view` and `jr issue comments` via `src/cli/issue/list.rs`) renders every `listItem` with a `- ` prefix, regardless of whether its parent is `bulletList` or `orderedList`. After #201 made `markdown_to_adf` emit proper `orderedList` nodes (with optional `attrs.order`), and after the write-path now supports many more ADF node types, the read-path renderer is behind.

Concrete user-visible effects:
- Ordered lists in Jira issues/comments display as bullet lists in `jr issue view`
- Tables render their content without structure (cells/rows collapse into a flat stream)
- Links lose their `href` attrs (only the link text shows)
- Bold/italic/strike/code marks are dropped
- Blockquotes have no marker to distinguish them from regular paragraphs
- Horizontal rules are silently dropped

## Scope (originally deferred from #197)

Flagged by Copilot during #201 review. The #197 spec explicitly placed `adf_to_text` reverse-direction support in "Out of Scope (Follow-Ups)" because the write-path replacement was the primary fix, and read-path is an orthogonal concern. Filing as a standalone issue so the fix can land on its own.

## Acceptance Criteria

- [ ] `orderedList` items render with numeric prefixes (`1. `, `2. `, ...), respecting `attrs.order` when present
- [ ] Nested lists keep correct indentation (2 spaces per level, already works for `bulletList`)
- [ ] `table` renders as a recognizable text representation (e.g., ASCII grid, pipes, or just rows with `|` separators)
- [ ] `link` mark preserves href: e.g., `[text](url)` or `text (url)` style
- [ ] Inline marks render visibly: `**bold**`, `*italic*`, `~~strike~~`, `` `code` ``
- [ ] `blockquote` prefixes each line with `> `
- [ ] `rule` renders as `---` on its own line
- [ ] Unit tests cover each case; snapshot test updates `test_adf_to_text_snapshot` with a rich input

## Related

- PR #201 — introduced the new write-path ADF shapes
- #197 — original feature that placed this in Out of Scope

### PR DESCRIPTION
**Pull request by Zious11**
## Summary

Replaces the read-path `adf_to_text` in `src/adf.rs` with a stateful `AdfRenderer` struct that fully covers the ADF node types the write-path (`markdown_to_adf` from #197) now emits. The renderer is symmetric to the existing `AdfBuilder` in the same file.

**New behavior:**
- Ordered lists render with numeric prefixes respecting `attrs.order` (values `< 1` fall back to `1` matching Jira's renderer)
- Inline marks render with markdown syntax: `**strong**`, `*em*`, `~~strike~~`, `` `code` ``, `[text](href)`; unknown mark types pass bare
- Blockquotes prefix each line with `> `; nested blockquotes accumulate to `> >` via stack-unwinding re-prefix
- `rule` → `---`, `hardBreak` → `\n`, `codeBlock` opens with `` ```{attrs.language} `` when present
- Tables render as pipe rows (`| cell | cell |`) with a `| --- | --- |` separator after any row containing a `tableHeader` cell
- Unknown container nodes recurse into `content`; unknown leaves drop silently (replaces the old `[unsupported: <type>]` debug string that leaked into `jr issue view`)

## Fixes

Closes #202

## Design & Plan

- Spec: `docs/superpowers/specs/2026-04-16-adf-to-text-rich-rendering-design.md`
- Plan: `docs/superpowers/plans/2026-04-16-adf-to-text-rich-rendering.md`
- 8 self-contained TDD commits (RED → GREEN → clippy-clean → commit per task)

## Test Plan

- [x] Unit tests: 55 adf tests passing (previously 32) — one per AC item plus edge cases (`order=0` defaults, mixed nested lists, unknown mark drop, link missing href, mixed header/cell table row, unknown leaf drops silently, etc.)
- [x] Snapshot test `adf_to_text_complex` regenerated with a rich ADF document exercising every new rendering path
- [x] Structural roundtrip test `test_markdown_to_adf_to_text_roundtrip` — markdown → ADF → text → ADF preserves node-type traversal order
- [x] `cargo fmt --all -- --check` clean
- [x] `cargo clippy --all-targets -- -D warnings` clean (zero lint suppressions)
- [x] Full `cargo test` passes (415 lib tests + integration tests)
- [x] Local multi-agent PR review: code-reviewer, pr-test-analyzer, silent-failure-hunter, type-design-analyzer, comment-analyzer — all clean after one round of comment-accuracy fixes
- [ ] Manual smoke: `jr issue view <KEY>` against live Jira on a description with rich formatting

## Follow-up

- #203 — expand edge-case test coverage (empty docs, whitespace-only blockquote, multi-hardBreak, more mark permutations). Non-critical; all deferred findings from the local review.
- Out-of-scope per the spec's Out of Scope section: `mention`, `status`, `emoji`, `mediaSingle`, `panel` type badge, `taskList/taskItem` — each a small follow-up when surfaced.

### CODE REVIEW

File: src/adf.rs [hunk 1]
```diff
@@ -343,70 +343,255 @@ fn heading_level_to_u8(level: HeadingLevel) -> u8 {
...
-        for node in content {
-            render_node(node, &mut output, 0);
+    let mut r = AdfRenderer::new();
+    r.render_doc(adf);
+    r.finish()
+}
+
+struct AdfRenderer {
+    output: String,
+    list_stack: Vec<ListFrame>,
```
[the remark this answers is not in the document]
Comment (Zious11) [reply src/adf.rs:353]: Fixed in 731a7ab. Removed the field entirely — the stack-unwind re-prefix pass already handles nested blockquote accumulation, so the counter was pure bookkeeping. Perplexity + local `cargo clippy --all-targets -- -D warnings` confirmed compound-assignment `self.field += 1` counts as a read for rustc's dead_code lint, so CI wasn't actually at risk, but removing the dead state is the cleaner fix.

File: src/adf.rs [hunk 2]
```diff
@@ -343,70 +343,255 @@ fn heading_level_to_u8(level: HeadingLevel) -> u8 {
...
+            "table" => {
+                self.render_children(node);
+                self.output.push('\n');
+            }
+            "tableRow" => {
+                let cells = node
+                    .get("content")
+                    .and_then(|c| c.as_array())
+                    .cloned()
+                    .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (Zious11) [reply src/adf.rs]: Fixed in 731a7ab. Switched to a slice borrow: `as_array().map(|v| v.as_slice()).unwrap_or(&[])`. Avoids the deep clone of cell `Value`s per row.

File: src/adf.rs [hunk 3]
```diff
@@ -343,70 +343,250 @@ fn heading_level_to_u8(level: HeadingLevel) -> u8 {
...
+/// Wrap `text` with markdown-style syntax for each mark, innermost-first.
+/// Unknown mark types pass through without added syntax.
+fn apply_marks(text: &str, marks: Option<&Vec<Value>>) -> String {
+    let mut result = text.to_string();
+    let Some(marks) = marks else { return result };
+    for mark in marks {
+        let mark_type = mark.get("type").and_then(|t| t.as_str()).unwrap_or("");
+        result = match mark_type {
+            "code" => format!("`{result}`"),
+            "em" => format!("*{result}*"),
```
[the remark this answers is not in the document]
Comment (Zious11) [reply src/adf.rs:667]: Fixed in fb21247. Added `wrap_code_span` helper that follows the CommonMark rule: delimiter length = longest backtick run in content + 1. Also pads with a single space when content starts/ends with a backtick so the delimiter can't glue to the content. New test `test_render_code_mark_with_backtick_in_content` covers `foo\`bar` → ````foo\`bar```` and `test_render_code_mark_with_leading_trailing_backtick_pads` covers the padding case.

File: src/adf.rs [hunk 4]
```diff
@@ -343,70 +343,250 @@ fn heading_level_to_u8(level: HeadingLevel) -> u8 {
...
+                // inner passes already produced — so a fixed "> " is correct at
+                // every level; no depth counter is needed.
+                let rendered = self.output.split_off(start);
+                let prefix = "> ";
+                for (i, line) in rendered.split('\n').enumerate() {
+                    if i > 0 {
+                        self.output.push('\n');
+                    }
+                    if !line.is_empty() {
+                        self.output.push_str(prefix);
```
[the remark this answers is not in the document]
Comment (Zious11) [reply src/adf.rs]: Fixed in fb21247. Blank lines inside a blockquote now get a `>` prefix (no trailing space, matching conventional markdown `>\n` form), trailing empties are trimmed. Preserves block-quote continuity when a nested node produces internal blank lines. New test `test_render_blockquote_with_internal_blank_line_keeps_prefix` covers a code block with blank lines inside a blockquote.

File: src/adf.rs [hunk 5]
```diff
@@ -343,70 +343,250 @@ fn heading_level_to_u8(level: HeadingLevel) -> u8 {
...
+            }
+            let child_type = child.get("type").and_then(|t| t.as_str()).unwrap_or("");
+            match child_type {
+                "paragraph" => {
+                    if let Some(cc) = child.get("content").and_then(|c| c.as_array()) {
+                        for inline in cc {
+                            self.render_node(inline);
+                        }
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (Zious11) [reply src/adf.rs:575]: Fixed in fb21247. Added a `render_inline_in_cell` helper that `render_cell_inline` uses for paragraph content: it substitutes a space for `hardBreak` so the pipe row stays on one line. Direct-`hardBreak` children at cell level are also handled. New test `test_render_hard_break_in_table_cell_becomes_space` covers it.

File: src/adf.rs [hunk 6]
```diff
@@ -343,70 +343,307 @@ fn heading_level_to_u8(level: HeadingLevel) -> u8 {
...
+/// Unknown mark types pass through without added syntax.
+fn apply_marks(text: &str, marks: Option<&Vec<Value>>) -> String {
+    let mut result = text.to_string();
+    let Some(marks) = marks else { return result };
+    for mark in marks {
+        let mark_type = mark.get("type").and_then(|t| t.as_str()).unwrap_or("");
+        result = match mark_type {
+            "code" => wrap_code_span(&result),
+            "em" => format!("*{result}*"),
+            "strong" => format!("**{result}**"),
```
[the remark this answers is not in the document]
Comment (Zious11) [reply src/adf.rs:668]: Fixed in 13d66e6. `apply_marks` now applies `code` innermost regardless of its position in the marks array. The fix is critical because the write-path's `push_code` (src/adf.rs:272-288) appends the code mark after active marks, so roundtrip yields `marks: [strong, code]` for `**` + inline code `**`. Under strict array-order iteration we were producing `` `**code**` `` (code outermost, semantics lost); now we produce `**` + `` `code` `` + `**` = ``**`code`**``. New test `test_render_strong_with_code_applies_code_innermost` covers it.

File: src/adf.rs [hunk 7]
```diff
@@ -343,70 +343,307 @@ fn heading_level_to_u8(level: HeadingLevel) -> u8 {
...
     }
+
+    /// Like `render_node`, but when the node is a `hardBreak` emit a space
+    /// instead of a newline. Used inside table cells to keep rows single-line.
+    fn render_inline_in_cell(&mut self, node: &Value) {
+        let t = node.get("type").and_then(|t| t.as_str()).unwrap_or("");
+        if t == "hardBreak" {
+            self.output.push(' ');
+        } else {
+            self.render_node(node);
```
[the remark this answers is not in the document]
Comment (Zious11) [reply src/adf.rs]: Fixed in 13d66e6. Added `sanitize_table_cell_text` helper (newline → space, `|` → `\|`) and `render_inline_in_cell` now dispatches text nodes through it, applying marks to the sanitized text so escapes survive `**foo|bar**` style input. New tests `test_render_table_cell_escapes_pipe_in_text` and `test_render_table_cell_collapses_newlines_in_text` cover both paths.

File: docs/superpowers/specs/2026-04-16-adf-to-text-rich-rendering-design.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+enum ListFrame {
+    Bullet,
+    Ordered { next_index: u64 },
+}
+```
+
+- `list_stack.len()` replaces the old `depth: usize` parameter. Indent level for `listItem` is `list_stack.len() - 1`.
+- Each `listItem` under `ListFrame::Ordered` increments `next_index` after rendering.
+- `blockquote_depth` tracks nested `blockquote` nesting; prefix rendered into each line as `"> " * blockquote_depth`.
```
[the remark this answers is not in the document]
Comment (Zious11) [reply docs/superpowers/specs/2026-04-16-adf-to-text-rich-rendering-design.md]: Fixed in 13d66e6. Updated the spec's `AdfRenderer` struct snippet, feature-mapping table row for `blockquote`, and "Blockquote line-prefixing" implementation section to reflect the stack-unwind re-prefix approach (no `blockquote_depth` field). The history of the field is preserved in the plan's self-review section for context.

File: docs/superpowers/plans/2026-04-16-adf-to-text-rich-rendering.md [hunk 9]
```diff
@@ -0,0 +1,1180 @@
...
+    r.render_doc(adf);
+    r.finish()
+}
+
+struct AdfRenderer {
+    output: String,
+    list_stack: Vec<ListFrame>,
+    blockquote_depth: usize,
+}
+
```
[the remark this answers is not in the document]
Comment (Zious11) [reply docs/superpowers/plans/2026-04-16-adf-to-text-rich-rendering.md:35]: Fixed in 13d66e6. Updated the plan's Architecture summary, Task 1 scaffold snippet (struct + `new()`), and Task 4 blockquote-arm snippet to match the final implementation. A trailing self-review bullet explicitly notes `blockquote_depth` was scoped out so the history is still visible.
